llvm/lib/Support/Unix/MemoryLinux.h

```diff
@@ -0,0 +1,109 @@
+//===- Unix/MemoryLinux.h - Linux specific Helper Fuctions ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines Linux specific helper functions for memory management.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_SUPPORT_UNIX_MEMORYLINUX_H
+#define LLVM_LIB_SUPPORT_UNIX_MEMORYLINUX_H
+
+#ifndef __linux__
+#error Linux only support header!
+#endif
+
+#include "llvm/Support/Process.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/mman.h>
+#include <sys/syscall.h>
+
+#ifndef MFD_CLOEXEC
+#define MFD_CLOEXEC 0x0001U
+#endif
+#ifndef MFD_EXEC
+#define MFD_EXEC 0x0010U
+#endif
+
+namespace llvm {
+namespace sys {
+namespace {
+
+static inline bool isPermissionError(int err) {
+  // PaX uses EPERM, SELinux uses EACCES
+  return err == EPERM || err == EACCES;
+}
+
+// FIXME: Make this either more low-level C'ish or C++'ish
+static inline bool execProtChangeNeedsNewMapping() {
```

Comment:
With this much implementation code, we'd probably want it in a .cpp file.

Reply:
My reasoning here was, as there's only one user for it (beside unit tests), it can live in a header file.

I had it in `llvm/lib/Support/Unix/Memory.inc` before but as the unit test `llvm/unittests/Support/MemoryTest.cpp` can make good use of it too, I moved the functionality to a header file so it can be shared.

But, really, this code should all be internal to `llvm/lib/Support/Unix/Memory.inc`, so there's no strong need for a cpp file (plus header for the declarations and #defines), imho.

Reply:
Ok sure. Mostly I was going off the base instinct that a lot of code in `.h` is usually not a good idea but if it's the same result in the end, fine.

Reply:
That instinct is usually right and I share it! Maybe I can think of a way to put it back to `llvm/lib/Support/Unix/Memory.inc` but still allow using it in the unit test.